src/Schema/Comparator.php

```diff
@@ -432,95 +434,76 @@ public function diffForeignKey(ForeignKeyConstraint $key1, ForeignKeyConstraint
     {
         $properties1 = $column1->toArray();
         $properties2 = $column2->toArray();
 
-        $changedProperties = [];
+        // Do not compare the column names
+        unset($properties1['name'], $properties2['name']);
 
-        if (get_class($properties1['type']) !== get_class($properties2['type'])) {
-            $changedProperties[] = 'type';
-        }
+        $changedProperties = [];
 
-        foreach (['notnull', 'unsigned', 'autoincrement'] as $property) {
-            if ($properties1[$property] === $properties2[$property]) {
+        foreach (array_unique(array_merge(array_keys($properties1), array_keys($properties2))) as $key) {
+            if (
+                array_key_exists($key, $properties1)
+                && array_key_exists($key, $properties2)
+                && $properties1[$key] === $properties2[$key]
+            ) {
                 continue;
             }
 
-            $changedProperties[] = $property;
-        }
-
-        // Null values need to be checked additionally as they tell whether to create or drop a default value.
-        // null != 0, null != false, null != '' etc. This affects platform's table alteration SQL generation.
-        if (
-            ($properties1['default'] === null) !== ($properties2['default'] === null)
-            || $properties1['default'] != $properties2['default']
-        ) {
-            $changedProperties[] = 'default';
+            $changedProperties[] = $key;
         }
 
-        if (
-            ($properties1['type'] instanceof Types\StringType && ! $properties1['type'] instanceof Types\GuidType) ||
-            $properties1['type'] instanceof Types\BinaryType
-        ) {
-            // check if value of length is set at all, default value assumed otherwise.
-            $length1 = $properties1['length'] ?? 255;
-            $length2 = $properties2['length'] ?? 255;
-            if ($length1 !== $length2) {
-                $changedProperties[] = 'length';
-            }
-
-            if ($properties1['fixed'] !== $properties2['fixed']) {
-                $changedProperties[] = 'fixed';
-            }
-        } elseif ($properties1['type'] instanceof Types\DecimalType) {
-            if (($properties1['precision'] ?? 10) !== ($properties2['precision'] ?? 10)) {
-                $changedProperties[] = 'precision';
+        return array_values(array_filter(
+            $changedProperties,
+            function ($property) use ($column1, $column2, $platform): bool {
+                return $this->changedPropertyHasEffect($property, $column1, $column2, $platform);
             }
+        ));
+    }
 
-            if ($properties1['scale'] !== $properties2['scale']) {
-                $changedProperties[] = 'scale';
+    /**
+     * Detects if a changed property has an actual effect on the used platform
+     */
+    private function changedPropertyHasEffect(
+        string $property,
```

Comment:
Why do we need to compare columns in the context of a single property? From the migration standpoint, we only need to know whether the columns are effectively equal or not. I believe we could remove some details from the resulting diff in order to not complicate the implementation.

Reply:
Makes sense. I switched to comparing the SQL only once per column in 5dbb71531556225f189f5bc1cd26009b762d0c84 and it seems to work fine.